File: go/test/notification_channel_setting_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/trello-sdk/go"
	"github.com/voxgig-sdk/trello-sdk/go/core"

	vs "github.com/voxgig-sdk/trello-sdk/go/utility/struct"
)

func TestNotificationChannelSettingEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.NotificationChannelSetting(nil)
		if ent == nil {
			t.Fatal("expected non-nil NotificationChannelSettingEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"notification_channel_setting": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.NotificationChannelSetting(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.NotificationChannelSetting(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := notification_channel_settingBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "notification_channel_setting." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		notificationChannelSettingRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.notification_channel_setting", setup.data)))
		var notificationChannelSettingRef01Data map[string]any
		if len(notificationChannelSettingRef01DataRaw) > 0 {
			notificationChannelSettingRef01Data = core.ToMapAny(notificationChannelSettingRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = notificationChannelSettingRef01Data

		// LIST
		notificationChannelSettingRef01Ent := client.NotificationChannelSetting(nil)
		notificationChannelSettingRef01Match := map[string]any{
			"member_id": setup.idmap["member01"],
		}

		notificationChannelSettingRef01ListResult, err := notificationChannelSettingRef01Ent.List(notificationChannelSettingRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, notificationChannelSettingRef01ListOk := notificationChannelSettingRef01ListResult.([]any)
		if !notificationChannelSettingRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", notificationChannelSettingRef01ListResult)
		}

		// UPDATE
		notificationChannelSettingRef01DataUp0Up := map[string]any{
			"id": notificationChannelSettingRef01Data["id"],
		}

		notificationChannelSettingRef01MarkdefUp0Name := "idMember"
		notificationChannelSettingRef01MarkdefUp0Value := fmt.Sprintf("Mark01-notification_channel_setting_ref01_%d", setup.now)
		notificationChannelSettingRef01DataUp0Up[notificationChannelSettingRef01MarkdefUp0Name] = notificationChannelSettingRef01MarkdefUp0Value

		notificationChannelSettingRef01ResdataUp0Result, err := notificationChannelSettingRef01Ent.Update(notificationChannelSettingRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		notificationChannelSettingRef01ResdataUp0 := core.ToMapAny(entityData(notificationChannelSettingRef01ResdataUp0Result))
		if notificationChannelSettingRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if notificationChannelSettingRef01ResdataUp0["id"] != notificationChannelSettingRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if notificationChannelSettingRef01ResdataUp0[notificationChannelSettingRef01MarkdefUp0Name] != notificationChannelSettingRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", notificationChannelSettingRef01MarkdefUp0Name, notificationChannelSettingRef01ResdataUp0[notificationChannelSettingRef01MarkdefUp0Name])
		}

		// LOAD
		notificationChannelSettingRef01MatchDt0 := map[string]any{
			"id": notificationChannelSettingRef01Data["id"],
		}
		notificationChannelSettingRef01DataDt0Loaded, err := notificationChannelSettingRef01Ent.Load(notificationChannelSettingRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		notificationChannelSettingRef01DataDt0LoadResult := core.ToMapAny(entityData(notificationChannelSettingRef01DataDt0Loaded))
		if notificationChannelSettingRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if notificationChannelSettingRef01DataDt0LoadResult["id"] != notificationChannelSettingRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func notification_channel_settingBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "notification_channel_setting", "NotificationChannelSettingTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read notification_channel_setting test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse notification_channel_setting test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"notification_channel_setting01", "notification_channel_setting02", "notification_channel_setting03", "member01", "member02", "member03", "notifications_channel_setting01", "notifications_channel_setting02", "notifications_channel_setting03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID": idmap,
		"TRELLO_TEST_LIVE":      "FALSE",
		"TRELLO_TEST_EXPLAIN":   "FALSE",
		"TRELLO_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["TRELLO_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["TRELLO_APIKEY"],
			},
			extra,
		})
		client = sdk.NewTrelloSDK(core.ToMapAny(mergedOpts))
	}

	live := env["TRELLO_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["TRELLO_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/NotificationChannelSettingEntityTest.php
<?php
declare(strict_types=1);

// NotificationChannelSetting entity test

require_once __DIR__ . '/../trello_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class NotificationChannelSettingEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TrelloSDK::test(null, null);
        $ent = $testsdk->NotificationChannelSetting(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "notification_channel_setting" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = TrelloSDK::test($seed, null);
        $seen = iterator_to_array($base->NotificationChannelSetting(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = TrelloConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = TrelloSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->NotificationChannelSetting(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = notification_channel_setting_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list", "update", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "notification_channel_setting." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $notification_channel_setting_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.notification_channel_setting")));
        $notification_channel_setting_ref01_data = null;
        if (count($notification_channel_setting_ref01_data_raw) > 0) {
            $notification_channel_setting_ref01_data = Helpers::to_map($notification_channel_setting_ref01_data_raw[0][1]);
        }

        // LIST
        $notification_channel_setting_ref01_ent = $client->NotificationChannelSetting(null);
        $notification_channel_setting_ref01_match = [
            "member_id" => $setup["idmap"]["member01"],
        ];

        $notification_channel_setting_ref01_list_result = $notification_channel_setting_ref01_ent->list($notification_channel_setting_ref01_match, null);
        $this->assertIsArray($notification_channel_setting_ref01_list_result);

        // UPDATE
        $notification_channel_setting_ref01_data_up0_up = [
            "id" => $notification_channel_setting_ref01_data["id"],
        ];

        $notification_channel_setting_ref01_markdef_up0_name = "idMember";
        $notification_channel_setting_ref01_markdef_up0_value = "Mark01-notification_channel_setting_ref01_" . $setup["now"];
        $notification_channel_setting_ref01_data_up0_up[$notification_channel_setting_ref01_markdef_up0_name] = $notification_channel_setting_ref01_markdef_up0_value;

        $notification_channel_setting_ref01_resdata_up0_result = $notification_channel_setting_ref01_ent->update($notification_channel_setting_ref01_data_up0_up, null);
        $notification_channel_setting_ref01_resdata_up0 = Helpers::to_map(is_object($notification_channel_setting_ref01_resdata_up0_result) && method_exists($notification_channel_setting_ref01_resdata_up0_result, 'data_get') ? $notification_channel_setting_ref01_resdata_up0_result->data_get() : $notification_channel_setting_ref01_resdata_up0_result);
        $this->assertNotNull($notification_channel_setting_ref01_resdata_up0);
        $this->assertEquals($notification_channel_setting_ref01_resdata_up0["id"], $notification_channel_setting_ref01_data_up0_up["id"]);
        $this->assertEquals($notification_channel_setting_ref01_resdata_up0[$notification_channel_setting_ref01_markdef_up0_name], $notification_channel_setting_ref01_markdef_up0_value);

        // LOAD
        $notification_channel_setting_ref01_match_dt0 = [
            "id" => $notification_channel_setting_ref01_data["id"],
        ];
        $notification_channel_setting_ref01_data_dt0_loaded = $notification_channel_setting_ref01_ent->load($notification_channel_setting_ref01_match_dt0, null);
        $notification_channel_setting_ref01_data_dt0_load_result = Helpers::to_map(is_object($notification_channel_setting_ref01_data_dt0_loaded) && method_exists($notification_channel_setting_ref01_data_dt0_loaded, 'data_get') ? $notification_channel_setting_ref01_data_dt0_loaded->data_get() : $notification_channel_setting_ref01_data_dt0_loaded);
        $this->assertNotNull($notification_channel_setting_ref01_data_dt0_load_result);
        $this->assertEquals($notification_channel_setting_ref01_data_dt0_load_result["id"], $notification_channel_setting_ref01_data["id"]);

    }
}

function notification_channel_setting_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/notification_channel_setting/NotificationChannelSettingTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TrelloSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["notification_channel_setting01", "notification_channel_setting02", "notification_channel_setting03", "member01", "member02", "member03", "notifications_channel_setting01", "notifications_channel_setting02", "notifications_channel_setting03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID" => $idmap,
        "TRELLO_TEST_LIVE" => "FALSE",
        "TRELLO_TEST_EXPLAIN" => "FALSE",
        "TRELLO_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TRELLO_TEST_NOTIFICATION_CHANNEL_SETTING_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TRELLO_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["TRELLO_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new TrelloSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TRELLO_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TRELLO_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
